core/src/apps/cardano/address.py

```diff
@@ -1,145 +1,344 @@
+    is_byron_address = parameters.address_type == CardanoAddressType.BYRON
+
+    if is_byron_address:
+        address = _derive_byron_address(
+            keychain, parameters.spending_key_path, protocol_magic
+        )
+    else:
+        address = _derive_shelley_address(keychain, parameters, network_id)
+
+    return address
 
-        if not isinstance(address_protocol_magic, int):
-            raise INVALID_ADDRESS
 
-        if address_protocol_magic != protocol_magic:
-            raise NETWORK_MISMATCH
+def _derive_byron_address(
+    keychain: seed.Keychain, path: List[int], protocol_magic: int
+) -> bytes:
+    if not is_byron_path(path):
+        raise wire.DataError("Invalid path for byron address!")
 
+    address = derive_byron_address(keychain, path, protocol_magic)
+    return address
 
-def validate_full_path(path: list) -> bool:
+
+def _derive_shelley_address(
+    keychain: seed.Keychain, parameters: CardanoAddressParametersType, network_id: int,
+) -> bytes:
+    if not is_shelley_path(parameters.spending_key_path):
+        raise wire.DataError("Invalid path for shelley address!")
+
+    address_header = _create_address_header(parameters.address_type, network_id)
+
+    if parameters.address_type == CardanoAddressType.BASE:
+        address = _derive_base_address(
+            keychain,
+            address_header,
+            parameters.spending_key_path,
+            parameters.staking_key_path,
+            parameters.staking_key_hash,
+        )
+    elif parameters.address_type == CardanoAddressType.ENTERPRISE:
+        address = _derive_enterprise_address(
+            keychain, address_header, parameters.spending_key_path
+        )
+    elif parameters.address_type == CardanoAddressType.POINTER:
+        address = _derive_pointer_address(
+            keychain,
+            address_header,
+            parameters.spending_key_path,
+            parameters.certificate_pointer,
+        )
+    elif parameters.address_type == CardanoAddressType.REWARD:
+        address = _derive_reward_address(
+            keychain, address_header, parameters.spending_key_path
+        )
+    else:
+        raise ValueError("Invalid address type '%s'!" % parameters.address_type)
+
+    return address
+
+
+def _create_address_header(address_type: int, network_id: int) -> bytes:
+    if address_type == CardanoAddressType.BYRON:
+        """
+        Byron addresses don't have an explicit header in the Shelley
+        spec. However, thanks to their CBOR structure they always start with
+        0b1000 - the byron address id. This is no coincidence.
+        The Shelley address headers are purposefully built around these
+        starting bits of the byron address.
+        """
+        raise ValueError("Byron address does not contain an explicit header!")
+
+    if not _validate_address_type(address_type):
+        raise ValueError("Invalid address type '%s'!" % address_type)
+
+    header = address_type << 4 | network_id
```

Comment:
reviewing this function ISTM you should drop all checks and reduce it to this and the below line

Reply:
Ok, removed [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/76bc4dcc3a298bc190d0cd4263fae7ed1d998d73).